### ISSUE TITLE
gcc-6.3.1 - actually pre-release

### ISSUE THREAD

**Issue by pullmoll** [voidlinux/void-packages#5971]
You may have noticed that Arch Linux updated their [gcc to 6.3.1](https://www.archlinux.org/packages/core/x86_64/gcc/), or actually a pre-release version with git commit hash 4fc407888a30c5d953816b05c8a8e98ec2ab3101.

This version does no longer have **java**
and it does **no** longer need a **patch set** for **musl libc**.

Since we eventually will also update to 6.3.1, we need to decide how to handle the java removal.
There is currently no package using gcc-gcj. The only pending use case was my try to build an openjdk-7 just for the purpose for bootstrapping our regular openjdk (8). Since this try still fails for x86_64-musl,  because of a problem with the boehm garbage collector and finding the correct memory range to scan, there is no good reason to wait. Edit: Correction - there is `pdftk` which uses gcc-gcj.

In case I would really find the time and courage to tackle this again, I could simply integrate gcc-gcj-6.3.0 into the pending openjdk7 package, building a local version of gcj for just that task.

There are, however, some more things to discuss. Do we ...
+ want to try to support gcc-ada? There has been one request for gcc-ada in the past ( #4703 ).
+ want to run gcc's fixincludes (as LFS does) or not (as Arch does and we did for gcc-4.x)?
+ want to keep libssp? Arch disabled it now and it seems no Void package actually uses it.
+ (optionally) run the `make -k check` which depends on dejagnu (not yet packaged) and inetutils?
+ still have the option to use `--with-default-libstdcxx-abi=gcc4-compatible`? Not => really bad!

My opinions on the questions:
ada: why not? it seems easy enough to add it.
fixincludes: unsure, because I haven't seen any pro or contra arguments.
libssp: we could keep it as it's no big thing.
check: pro; could be done by merging #5813 and adding a `do_check()` to the gcc/template.
gcc4-compat: we don't want to rebuild everything, do we?

Another interesting thing to consider is Alpine's way of enabling hardening (strong stack protection, fortify-source, etc.) as **default**. This could get us rid of the (still) problematic `-specs=...` and modified C{,XX}FLAGS, which are leaking into pkgconfig files and the like, if not patched away. The unusual `nopie=yes` case would then have to modify the C{,XX}FLAGS to contain `-nopie`.
See the [patchset at alpinelinux.git](http://git.alpinelinux.org/cgit/aports/tree/main/gcc?h=master) for their way to deal with it.

I'm willing to do the usual bootstrap and build of basic packages to make sure it works, beforehand, once the details are decided upon. We should probably create a local tarball again instead of git cloning gcc.

**Comment by chneukirchen** [comment]
ada: ok to add
fixincludes: leave as is
libssp: leave as is
check: ok, with do_check
gcc4-compat: keep, but it's time we tackle this one
gcj: remove it, drop pdftk (steal pdftk-bin from debian if you have to)
nopie: could be interesting, but needs more testing i guess

That said, what's the time frame for gcc 6.4? I'd prefer we stick to official releases.

**Comment by pullmoll** [comment]
I've seen no planned release date for 6.4. Judging by the usual rate (~ 4 months per minor release cycle) it could be around april or may this year.

I could give 6.3.1-pre a try without the intention to merge it, just to see how it behaves and to solve problems on the walk through.

**Comment by pullmoll** [comment]
Repository https://github.com/pullmoll/void-packages-gcc631/
Packages http://repo.voidlinux.de/current/
Ada will be a separate package (gcc-ada) because integration into gcc is too hairy for my taste.
I will upload live images as soon as sufficient packages are built to create them.

**Comment by pullmoll** [comment]
Superseeded by gcc-7.2.0 now.

**Comment by avoidr** [comment]
>want to try to support gcc-ada? There has been one request for gcc-ada in the past ( #4703 ).

Second.

>ada: why not? it seems easy enough to add it.

Please do! I need it, too, for GHDL.

**Comment by 1ma** [comment]
What's the status of gcc-ada?

**Comment by pullmoll** [comment]
I recently tried to build it but failed. My template to build gcc-ada as a separate package [is here](https://gist.github.com/pullmoll/df14ce845410df1a97fe2793e2d87a12) in case someone wants to pick up and continue.

**Comment by 1ma** [comment]
Ok I'm totally clueless about building void packages in general and GCC in particular but here's what I did, and what I've found:

```bash
$ git clone https://github.com/void-linux/void-packages
$ cd void-packages
$ mkdir srcpkgs/gcc-ada
$ curl https://gist.githubusercontent.com/pullmoll/df14ce845410df1a97fe2793e2d87a12/raw/797ed4a9a759ec497e67e7441b4c1c26280eab74/template > srcpkgs/gcc-ada/template
$ ./xbps-src binary-bootstrap
$ ./xbps-src pkg gcc-ada
```

...and the exact moment the build croaks [(full build log)](https://gist.github.com/1ma/e80dd3469a16fd05fa528b87b42edcc8):

```
(cd ada/bldtools/sinfo; gnatmake -q xsinfo ; ./xsinfo sinfo.h )
/bin/sh: gnatmake: command not found
/bin/sh: ./xsinfo: No such file or directory
make[3]: *** [../../gcc/ada/Make-generated.in:45: ada/sinfo.h] Error 127
make[3]: Leaving directory '/builddir/gcc-7.3.0/build/gcc'
make[2]: *** [Makefile:4548: all-stage1-gcc] Error 2
make[2]: Leaving directory '/builddir/gcc-7.3.0/build'
make[1]: *** [Makefile:22630: stage1-bubble] Error 2
make[1]: Leaving directory '/builddir/gcc-7.3.0/build'
make: *** [Makefile:934: all] Error 2
=> ERROR: gcc-ada-7.3.0_1: do_build: 'make ${makejobs}' exited with 2
=> ERROR:   in do_build() at srcpkgs/gcc-ada/template:98
```

According to the [GCC building instructions (Building the Ada compiler)](https://gcc.gnu.org/install/build.html), in order to build GNAT you need a preexisting Ada compiler, I guess that's why your template pulls down and installs AdaCore's GNAT. However, even though `gnatmake` and related tools are installed and available at `masterdir/builddir/gcc-7.3.0/gnat/bin`, the build process doesn't see them. How could we fix this?


PS Tell me if I should rather move that convo to another issue in the new repo.
